### PR TITLE
RSPEC-1186 Methods should not be empty

### LINKED ISSUE #66: Rule S1186: Methods should not be empty
Implements [RSPEC-1186](https://jira.sonarsource.com/browse/RSPEC-1186)

### PR DESCRIPTION
**Pull request by stylianos-agapiou-sonarsource**
fixes #66

### CODE REVIEW

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/EmptyMethodsCheck.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    Optional<UastNode> block;
+    if (node.is(UastNode.Kind.STATEMENT)) {
+      block = returnBlockFromGoStatement(node);
+    } else {
+      block = node.getChild(UastNode.Kind.BLOCK);
+    }
+    block.ifPresent(x -> {
+      if (x.children.stream().allMatch(child -> isBraceBracket(child.token))) {
+        reportIssue(node,
+          "Add a nested comment explaining why this method is empty, throw an [UnsupportedOperationException|NotSupportedException|IllegalOperationError] or complete the implementation.");
```
Comment (SonarTech) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/EmptyMethodsCheck.java]: Code Smell Code Smell: Split this 188 characters long line (which is greater than 180 authorized). ([squid:S00103](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=274&issues=AWL3qOSDqJ_F9Z_sg_cn&open=AWL3qOSDqJ_F9Z_sg_cn)

[ISSUE_KEY]: AWL3qOSDqJ_F9Z_sg_cn

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/EmptyMethodsCheck.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        if (funcLiteral.isPresent()) {
+          return funcLiteral.get().getChild(UastNode.Kind.BLOCK);
+        }
+      }
+    }
+    return Optional.empty();
+  }
+
+  private static boolean isBraceBracket(UastNode.Token token) {
+    return token != null && ("{".equals(token.value) || "}".equals(token.value));
```
Comment (SonarTech) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/EmptyMethodsCheck.java]: Code Smell Code Smell: Remove this expression which always evaluates to "true" ([squid:S2589](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2589))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=274&issues=AWL3qOSDqJ_F9Z_sg_cm&open=AWL3qOSDqJ_F9Z_sg_cm)

[ISSUE_KEY]: AWL3qOSDqJ_F9Z_sg_cm

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/EmptyMethodsCheck.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.util.Optional;
+import javax.annotation.Nullable;
+import org.sonar.check.Rule;
+import org.sonar.uast.UastNode;
+
+@Rule(key = "S1186")
+public class EmptyMethodsCheck extends Check {
+
+  public EmptyMethodsCheck() {
```
Comment (vilchik-elena) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/EmptyMethodsCheck.java]: @stylianos-agapiou-sonarsource rename class to `EmptyFunctionCheck`

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/EmptyMethodsCheck.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    super(UastNode.Kind.FUNCTION, UastNode.Kind.FUNCTION_LITERAL);
+  }
+
+  @Override
+  public void visitNode(UastNode node) {
+    Optional<UastNode> block;
+    block = node.getChild(UastNode.Kind.BLOCK);
```
Comment (vilchik-elena) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/EmptyMethodsCheck.java]: @stylianos-agapiou-sonarsource merge with declaration

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/EmptyMethodsCheck.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    block.ifPresent(blck -> {
+      if (blck.children.stream().allMatch(child -> isBraceBracket(child.token))) {
+        reportIssue(node,
+          "Add a nested comment explaining why this method is empty, throw an [UnsupportedOperationException|"
```
Comment (vilchik-elena) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/EmptyMethodsCheck.java]: @stylianos-agapiou-sonarsource such messages in RSPEC need edit (there are options for different languages). Looks like in Go there is no alternative, so I would just drop it: `Add a nested comment explaining why this function is empty or complete the implementation.`

File: common-rule-engine/src/test/files/checks/EmptyMethodsCheck/EmptyMethodsCheck.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
+package EmptyMethodsCheck
+
+import "fmt"
+
+func bar(a, b int) string{
+	return ""
+}
+// Noncompliant@+1 {{Add a nested comment explaining why this method is empty, throw an [UnsupportedOperationException|NotSupportedException|IllegalOperationError] or complete the implementation.}}
```
Comment (vilchik-elena) [inline common-rule-engine/src/test/files/checks/EmptyMethodsCheck/EmptyMethodsCheck.go]: @stylianos-agapiou-sonarsource assert primary location at least once

File: common-rule-engine/src/test/files/checks/EmptyMethodsCheck/EmptyMethodsCheck.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+	/* anonymous functions */
+	f := func(x, y int) int{return x+y}  // Compliant
+	f(1,2)
+
+	// Noncompliant@+1
+	f1 := func(x,y int) {}
+	f1(1,2)
+}
+
+// same rule works for methods and functions (methods' examples below)
```
Comment (vilchik-elena) [inline common-rule-engine/src/test/files/checks/EmptyMethodsCheck/EmptyMethodsCheck.go]: @stylianos-agapiou-sonarsource it's not `examples` :) it's tests
Avoid such story-telling comments, be concise and precise. 
IMO I would just drop this comment, it doesn't bring new info

File: common-rule-engine/src/test/files/checks/EmptyMethodsCheck/EmptyMethodsCheck.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+type Point struct {
+	x, y uint64
+}
+
+// Noncompliant@+1
+func(p *Point) Cycle(factor uint64){
+}
+
+func(p *Point)Length(){
+	// computes the length
```
Comment (vilchik-elena) [inline common-rule-engine/src/test/files/checks/EmptyMethodsCheck/EmptyMethodsCheck.go]: @stylianos-agapiou-sonarsource it's better to have `// compliant with comment inside` or smth like this, then it's easier for maintainers to understand the purpose of this case

File: sonar-go-plugin/src/main/resources/org/sonar/l10n/go/rules/go/S1186.json [hunk 9]
```diff
@@ -0,0 +1,16 @@
+{
+  "title": "Methods should not be empty",
```
Comment (vilchik-elena) [inline sonar-go-plugin/src/main/resources/org/sonar/l10n/go/rules/go/S1186.json]: @stylianos-agapiou-sonarsource change title in RSPEC to `Functions` and regenerate this file

File: its/ruling/src/test/expected/go-S1186.json [hunk 10]
```diff
@@ -0,0 +1,1321 @@
...
+126,
+127,
+],
+'project:kubernetes/pkg/kubectl/cmd/testing/fake.go':[
+211,
+448,
+455,
+458,
+],
+'project:kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime/api.pb.go':[
```
Comment (vilchik-elena) [inline its/ruling/src/test/expected/go-S1186.json]: @stylianos-agapiou-sonarsource this file is generated, it's better to exclude it from ruling (let's do it together tomorrow)

File: its/ruling/src/test/expected/go-S1186.json [hunk 11]
```diff
@@ -0,0 +1,1321 @@
...
+145,
+156,
+],
+'project:kubernetes/vendor/github.com/coreos/etcd/etcdserver/etcdserverpb/raft_internal.pb.go':[
+32,
+69,
+77,
+92,
+],
+'project:kubernetes/vendor/github.com/coreos/etcd/etcdserver/etcdserverpb/rpc.pb.go':[
```
Comment (vilchik-elena) [inline its/ruling/src/test/expected/go-S1186.json]: @stylianos-agapiou-sonarsource same here, generated

File: its/ruling/src/test/expected/go-S1186.json [hunk 12]
```diff
@@ -0,0 +1,1321 @@
...
+47,
+58,
+],
+'project:kubernetes/vendor/github.com/coreos/pkg/capnslog/journald_formatter.go':[
+68,
+],
+'project:kubernetes/vendor/github.com/coreos/pkg/capnslog/syslog_formatter.go':[
+64,
+],
+'project:kubernetes/vendor/github.com/coreos/rkt/api/v1alpha/api.pb.go':[
```
Comment (vilchik-elena) [inline its/ruling/src/test/expected/go-S1186.json]: @stylianos-agapiou-sonarsource and here

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/EmptyFunctionCheck.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    super(UastNode.Kind.FUNCTION, UastNode.Kind.FUNCTION_LITERAL);
+  }
+
+  @Override
+  public void visitNode(UastNode node) {
+    Optional<UastNode> block = node.getChild(UastNode.Kind.BLOCK);
+    block.ifPresent(blck -> {
+      if (blck.children.stream().allMatch(child -> isBraceBracket(child.token))) {
+        reportIssue(node,
+          "Add a nested comment explaining why this method is empty or complete the implementation.");
```
Comment (vilchik-elena) [inline common-rule-engine/src/main/java/org/sonar/commonruleengine/checks/EmptyFunctionCheck.java]: @stylianos-agapiou-sonarsource method -> function

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

**Review by vilchik-elena — APPROVED** [review]

**Pull request merged**